### ISSUE TITLE
load-languages command adds ~1 sec to startup time

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#2874]
The load-languages command is used when OpenRefine starts up, even though the information isn't shown to the user anywhere (unless they switch the Languages tab). It takes almost a full second (980 ms) to complete because the server is loading every single language file, just to get the language's name.

https://github.com/OpenRefine/OpenRefine/blob/10c046698a9586662668e31086bdd2d3c6aad801/main/src/com/google/refine/commands/lang/GetLanguagesCommand.java#L75-L82

The short term fix is to defer loading the language list until it's needed to render the Languages tab.

Longer term we can investigate caching the list on the server and/or initializing it at server startup time.

**Comment by tfmorris** [comment]
After looking at the code, I'm not sure this is actually blocking the UI, so I need to double check the performance graph.

**Comment by wetneb** [comment]
I tried to reproduce this: on my side it takes 30ms. I am using an SSD so perhaps that makes a big difference for accessing multiple files? @elroykanye just tried with an HDD and got 50 ms. So I am not sure what makes it so slow on your side @tfmorris. If you could check if you can still reproduce this two years after, it would be awesome.

**Comment by tfmorris** [comment]
For a cold start (OpenRefine 3.5.2 standard release server running, but not used since yesterday), I see several load-language requests with latencies as follows:

4120 msec
1440 msec
 944 msec
1200 msec

Overall the page took almost a minute to load:
59 requests 2.2 MB transferred 2.2 MB resources Finish: 1.0 min DOMContentLoaded: 50.78 s Load: 58.19 s

A warm refresh of the screen produced `load-language` latencies of 388, 39, 546, 376 msec (along with a  94 ms `get-preference?name=reconciliation.standardServices` call) for a total latency penalty of about 1.5 seconds.

Total home page refresh clocked in at a little over 18 seconds:
60 requests 2.2 MB transferred 2.2 MB resources Finish: 17.77 s DOMContentLoaded: 14.87 s Load: 17.39 s

This is all with Chrome Version 103.0.5060.114 (Official Build) (x86_64) running on macOS 11.6.2, 1 TB PCIe SSD disk, 2.5 GHz Core i7 processor, 16 GB DDR3 RAM (ie not the latest and greatest, but certainly a more than capable machine).

Here are the HAR files for the two loads:
[OpenRefine-3.5.2-Mac-Startup.har.zip](https://github.com/OpenRefine/OpenRefine/files/9152188/OpenRefine-3.5.2-Mac-Startup.har.zip)
[OpenRefine-3.5.2-Mac-Home-Refresh.har.zip](https://github.com/OpenRefine/OpenRefine/files/9152190/OpenRefine-3.5.2-Mac-Home-Refresh.har.zip)

**Comment by tfmorris** [comment]
There are a couple of independent issues:
* All the calls are synchronous and many are serialized unnecessarily (I think there's a separate issue for this)
* There are multiple `load-language` calls (although it's probably one per extension, which makes sense from a modularity point of view and wouldn't be an issue if they were asynchronous)
* The individual calls are slow, which was the original problem report (although, again, much less of an issue if asynchronous)

**Comment by wetneb** [comment]
Thanks a lot! I am surprised by the timings, they are indeed very long on your side although your machine is much better than mine in all respects. That's very curious!

Note that this issue is originally about the `get-languages` command (fired only once and only from the index page), not the `load-language` command (fired once per module, on the index and project pages).

**Comment by thadguidry** [comment]
Are the Java environments the same from both of you?

**Comment by tfmorris** [comment]
> Note that this issue is originally about the `get-languages` command (fired only once and only from the index page), not the `load-language` command (fired once per module, on the index and project pages).

Oops! `get-languages?&name=module&value=core` latency was 4560 msec on cold start and 294 msec on refresh.

My JRE is whatever's bundled with the binary Mac distribution for OpenRefine.

It's possible that an anti-virus program or something else is adding latency, but making things lazy loaded and async would make that irrelevant.
